src/dmd/e2ir.d

```diff
@@ -268,6 +268,43 @@ private elem *callfunc(const ref Loc loc,
                  */
                 ea.Ety = TYllong;
             }
+
+            /* Do integral promotions. This is not necessary per the C ABI, but
+             * some code from the C world seems to rely on it.
```

Comment:
This description is a bit misleading. Integral promotion is a language-level concept, specified by the language spec. The use of zero/sign-extension  when passing arguments in registers is something specified by the platform-specific ABI (psABI). Of course, the two are closely related. Still, this mixes concepts that describe things at different levels, and which don't always map cleanly between each other. For instance, the RISC-V psABI does zero/sign-extension to 32 bits and then sign-extension to 64 bits (for RV64), which is something that would be surprising when reasoning in terms of integral promotions.

Anyway, I just quickly checked the latest version of the x86-64 psABI and it seems that it's still omissive about the issue addressed by this PR. That's quite surprising, as this is one of the fundamental things that a psABI specifies. The only relevant thing I found was a footnote about `_Bool`, saying that bits 7-1 are specified to be 0, but the others are unspecified. That is consistent with the general rule about the unused bits being unspecified. [Microsoft's x86-64 ABI document][1] is more explicit about this and says that "All integer arguments in registers are right-justified, so the callee can ignore the upper bits of the register and access only the portion of the register necessary". (Even that description is ambiguous, as it's still consistent with the upper bits being specified by some rule).

I will try to gather more information about how this is currently being handled in the real-world by various compilers. In any case, I suppose there's no major cost for this "fix", even if this has been addressed by more recent versions of the compilers and the curses library, so it probably makes sense to merge it. I just recommend rewording the comment.

[1]: https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019#parameter-passing

Reply:
Thank you for doing the research on this. Unfortunately, we can't tell people to fix their C code, they'll just say "D is broken". We simply have to generate code that will work with their C code, regardless of what the spec says.

Reply:
> Unfortunately, we can't tell people to fix their C code, they'll just say "D is broken". We simply have to generate code that will work with their C code, regardless of what the spec says.

I agree. I was just wondering if the situation on the ground could have changed in the meanwhile. The relevant compilers could have been updated and newer versions of the libraries distributed, such that this no longer was an actual issue for reasonably recent distributions. I'll delve into this later. It seems that, at least for a while, some major compilers were disagreeing about this issue.

Reply:
I didn't have time to check more deeply but it seems that GCC is conservative both on the caller side (clearing out the upper bits before calling) and the callee side (zero/sign-extending before using the arguments), while Clang is conservative on the caller side but optimistically assumes that they were already extended when using them on the callee side. I tested with versions of GCC and Clang from trunk to quite old versions, and that remained the same. This is consistent with the bug report, which was reported for OS X, so the library was probably compiled with Clang. The curses library is incorrectly implementing a function with an `int` parameter that is declared as a `short` in the prototype. In theory that's undefined behavior, in practice it exposes these differences in how the ABI is handled by the various compilers. I agree that the sensible solution is for DMD to also be conservative on the caller side. On the callee side we're probably safe, given that Clang has lived with its approach for a long while.

Reply:
> The curses library is incorrectly implementing a function with an int parameter that is declared as a short in the prototype.

I can't see that in the code I posted, what I see is (after expanding macros)
```
int _nc_init_pair(SCREEN *sp, int pair, int f, int b)
{
   // ...
}

int init_pair_sp (SCREEN *sp, short pair, short f, short b)
{
  return _nc_init_pair(sp, pair, f, b);
}

int init_pair (short pair, short f, short b)
{
   return init_pair_sp(CURRENT_SCREEN, pair, f, b);
}
```

Reply:
> I can't see that in the code I posted, what I see is (after expanding macros)

Thanks. I was writing that bit from memory, and I probably also misdiagnosed that when I last looked at it. That further cements that we need this fix.